iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def validate(self) -> None:
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "season", "elapsed", "interval", "estimate", "previous"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert all(
+            self[x].dtype == pl.Utf8 for x in ["region", "season"]
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric()
+            for x in ["elapsed", "interval", "estimate", "previous"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class IncidentUptakeProjection(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "interval", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert (
+            self["region"].dtype == pl.Utf8
+        ), "Column 'region' and/or 'season' is not a string."
+        assert all(
+            self[x].dtype.is_numeric() for x in ["interval", "estimate"]
+        ), "Column 'elapsed', 'interval', 'estimate' and/or 'previous' is not numeric."
+
+
+class CumulativeUptakeData(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self, rollout: dt.datetime) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+                estimate=pl.col("estimate")
+                .diff()
+                .fill_null(pl.col("estimate").first()),
+                elapsed=(pl.col("date") - rollout).cast(pl.Float64)
+                / (1000 * 60 * 60 * 24),
```

Comment:
We want time units of days, which doesn't appear to be an option for a Duration object (available units are micro-, nano-, or milliseconds ... ?). I will search for an alternative way to turn "elapsed" into a number of days.

Reply:
Oh, right, "day" isn't a physical time duration, since it it is usually 24 hours but can be 23 or 25, and maybe something else with leap second too

Then I think the thing to use it `total_days()`?

Reply:
That's it! Andy just pointed me to [total_days](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.dt.total_days.html) also. Thanks!